source/extensions/tracers/xray/daemon_broker.cc

```diff
@@ -0,0 +1,51 @@
+#include "extensions/tracers/xray/daemon_broker.h"
+
+#include "common/network/utility.h"
+
+#include "source/extensions/tracers/xray/daemon.pb.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Tracers {
+namespace XRay {
+
+namespace {
+// creates a header JSON for X-Ray daemon.
+// For example:
+// { "format": "json", "version": 1}
+std::string createHeader(const std::string& format, uint32_t version) {
+  using namespace envoy::tracers::xray;
+  daemon::Header header;
+  header.set_format(format);
+  header.set_version(version);
+
+  Protobuf::util::JsonPrintOptions json_options;
+  json_options.preserve_proto_field_names = true;
+  std::string json;
+  const auto status = Protobuf::util::MessageToJsonString(header, &json, json_options);
+  if (!status.ok()) {
+    throw new EnvoyException("Failed to serialize X-Ray header to JSON.");
+  }
+  return json;
+}
+
+} // namespace
+
+DaemonBrokerImpl::DaemonBrokerImpl(const std::string& daemon_endpoint) {
+  using Network::Utility;
+  const auto address = Utility::parseInternetAddressAndPort(daemon_endpoint, false /*v6only*/);
+  io_handle_ = address->socket(Network::Address::SocketType::Datagram);
+  ASSERT(io_handle_->fd() != -1);
+  const auto result = address->connect(io_handle_->fd());
+  ASSERT(result.rc_ != -1);
```

Comment:
Why do we assert here instead of logging an error and failing in a more graceful manner?

Reply:
yeah - I don't know why I kept putting that off.
I need to add logging and not call connect if I don't have a valid descriptor at least.

I don't want to throw in this case because it's not a critical thing if we can't send the traces.. WDYT?

Reply:
yeah logging for sure.